src/types/response.ts

```diff
@@ -1,9 +1,28 @@
-export type TypedResponse<data> = {
-  data: data
-  format: 'cast-action' | 'frame' | 'transaction'
-}
+import type { ErrorFrameResponse } from './frame.js'
 
-export type HandlerResponse<typedResponse> =
+export type TypedResponse<
+  data,
+  format extends 'cast-action' | 'frame' | 'transaction',
+> = format extends 'cast-action' | 'transaction'
+  ? {
+      data: data
+      format: format
+    }
+  : { format: format } & (
+      | {
+          isErrorResponse: true
+          error: ErrorFrameResponse
+        }
+      | {
+          isErrorResponse: false
+          data: data
+        }
+    )
```

Comment:
I wonder if it's worth going down the route of minimal abstraction for now until other variants of errors are introduced:

https://www.typescriptlang.org/play?ssl=17&ssc=59&pln=17&pc=66#code/C4TwDgpgBAQghgZwgUQE6oPaqgXigbygFsIEE4BzCALigWFQEsA7CgGjuDmAFcEBhDABMIAflrMeRAEYRsAXwBQi0JCgAVcBCEAlUmAzMkAHiHc4APlwFFUKADMsRbrQDkAY0TAAtHHfBGQ1coAB8oV3tUOBJgsNcGOCM-AKDFeSgAMigAeWYIbPtjQjMuWhK4DnpuPjcEHnd3UgRg9LDCOUxUWngkNE7Krl4ENw6sFotld0N6KFR9aYgAfUjo6DxJABsNqEQoHmYAa2YMAHdmHYQNLV15owhjeiZWCcU5hAM75aiSADpyxQA9AC7CDQaCAHqiV63JBfVY-UaoQHAsGoqCQ5RvD6wla-KpDXA4PCuOoNJrBDJZLELOG-f5AtGMpnMllojHQ9403EQH74viE4mIilUmFLbkI9BYH4kMiUCDI1mKpXKuzsoHqjWarXanW6vU65SqaDIABucA2PG49yNVjwhAA2gcICAoCwoE6QBh7FAjQBdWhGx3O31QdJZfZHU7MQ1aHJ5ArGWx7ZiBc4QAAewAgzCElww0gAVhB-Gwk+qkx7LhmsznLgBpZ1egCqKcMxn2qdtUAbnvsLdT7dbzAspa7HcMUGr2dzruY9jkUAAklmiEnRFBTebLVnjMuIERMgQoPa67ON+n3BaRMZKxwPV6lyuLL7xMmRPYWNpQxM7BIICa5BUWMe2bIdjBtawjUnTNp0uKD13vb0oL-ADUCAA

Reply:
I think it's cool to have it fully abstracted – meaning that the `error` type would exists at all response formats.

On top of it – cast actions have exactly same API for errors, so that's +1 use case for going for maximum abstraction.
Yet I hope they add the same for transaction routes too, we can hide it from the user in the meantime and place a comment that it's not yet available.